pkg/controllers/nodeclass/status/nodeimage.go

```diff
@@ -0,0 +1,193 @@
+		reqs := lo.Map(nodeImage.Requirements.NodeSelectorRequirements(), func(item v1.NodeSelectorRequirementWithMinValues, _ int) corev1.NodeSelectorRequirement {
+			return item.NodeSelectorRequirement
+		})
+
+		sort.Slice(reqs, func(i, j int) bool {
+			if len(reqs[i].Key) != len(reqs[j].Key) {
+				return len(reqs[i].Key) < len(reqs[j].Key)
+			}
+			return reqs[i].Key < reqs[j].Key
+		})
+		return v1alpha2.NodeImage{
+			ID:           nodeImage.ID,
+			Requirements: reqs,
+		}
+	})
+
+	// Scenario A: Check if we should do a full update to latest before processing any partial update
+	shouldUpdate := imageVersionsUnready(nodeClass) || isMaintenanceWindowOpen()
+	logger.Debugf("should complete a full update to latest: %t", shouldUpdate)
+	if !shouldUpdate {
+		// Scenario B: Check if we should do any partial update based on image selectors, or newly supports SKUs
+		images, shouldUpdate = processPartialUpdate(nodeClass, images)
+	}
+
+	logger.Debugf("should update overall: %t", shouldUpdate)
+	if shouldUpdate {
+		if len(images) == 0 {
+			nodeClass.Status.NodeImages = nil
+			nodeClass.StatusConditions().SetFalse(v1alpha2.ConditionTypeNodeImageReady, "NodeImagesNotFound", "NodeImageSelectors did not match any NodeImages")
+			logger.Info("no images")
+			return reconcile.Result{RequeueAfter: time.Minute}, nil
+		}
+
+		nodeClass.Status.NodeImages = images
+		nodeClass.StatusConditions().SetTrue(v1alpha2.ConditionTypeNodeImageReady)
+	}
+
+	logger.Debug("success")
+	return reconcile.Result{RequeueAfter: 5 * time.Minute}, nil
+}
+
+// Case 1: This is a new AKSNodeClass, where node images haven't been populated yet
+// Case 2: This is indirectly handling k8s version image bump, since k8s version sets this status to false
+// Case 3: Note: like k8s we would also indirectly handle node features that required an image version bump, but none required atm.
+func imageVersionsUnready(nodeClass *v1alpha2.AKSNodeClass) bool {
+	return !nodeClass.StatusConditions().Get(v1alpha2.ConditionTypeNodeImageReady).IsTrue()
+}
+
+func isMaintenanceWindowOpen() bool {
+	// Case 4: check if the maintenance window is open
+	// TODO: need to add the actual logic for handling maintenance windows once it is in the ConfigMap.
+	return true
+}
+
+// Case 5: Handles the case of users updating image selectors.
+//   - Currently, this is just image family, and/or usage of SIG, which means that we should just be looking at the baseID of the images
+//
+// Case 6: We will softly add newly supported SKUs by Karpenter on their latest version
+//   - Note: I think this should be assess, if this is the exact behavior we want to give users, before any actual new SKU support is released.
+//
+// TODO: Need longer term design for handling newly supported versions, and other image selectors.
+func processPartialUpdate(nodeClass *v1alpha2.AKSNodeClass, discoveredImages []v1alpha2.NodeImage) ([]v1alpha2.NodeImage, bool) {
```

Comment:
defining what a partial update is at the top of this function comment, would I think be helpful

Reply:
Renamed method, and left a comment explaining it's logic